File: Algorithms/1. Sorting and Selection/selection-sort/selectionSort.cpp
// how to run: c++ -std=c++11 file.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
  cout << "\nSelection Sort: O(n^2) algorithm\n";

  // declare a vector
  vector<int> array = {5, 4, 3, 2, 1};

  // array before the sort
  cout << "Array before the sort: ";
  for (auto item: array)
    cout << item << " ";
  cout << endl;

  // run selection sort
  for (int i = 0; i < array.size(); i++) {
    int minIndex = i; // record current location
    for (int j = i+1; j < array.size(); j++) {
      // if current value is not minimum, set minIndex to j
      if (array[minIndex] > array[j])
        minIndex = j;
    }
    // swap the current ith value with the minIndex value
    swap(array[i], array[minIndex]);
  }

  // array after the sort
  cout << "Array after the sort: ";
  for (auto item: array)
    cout << item << " ";
  cout << endl;
}

File: Algorithms Techniques/1. Sorting and Selection/selection-sort/selectionSort.py
print("Selection Sort: O(n^2) Algorithm")

# create an array for testing
array = [64, 25, 12, 22, 11]

# array before sorting
print("Array before sorting:")
print(array)

# run selection sort
for i in range(len(array)):
    # Find the minimum element in remain unsorted array
    min_idx = i
    for j in range(i+1, len(array)):
        if array[min_idx] > array[j]: # current location is not minimum
            min_idx = j

    # Swap the found minimum element with current ith element
    array[i], array[min_idx] = array[min_idx], array[i]

# array after sorting
print("Array after sorting:")
print(array)
